### PR TITLE
Core/Currency: Several fixes and improvements

### PR DESCRIPTION
**Pull request by meji46**
**Changes proposed:**

- Defined several enums related to currencies
- Drop obsolete currencies and max caps config entries
- Fixed SMSG_SET_CURRENCY
- Added support for Quantity Gain/Lost Source for SMSG_SET_CURRENCY. Added helpers AddCurrency & RemoveCurrency that use these enums. Problems with weekly cap limitations for refund gain sources should also be fixed
- Removed incorrect restriction for not sending currencies with category CURRENCY_CATEGORY_META_CONQUEST in SMSG_SETUP_CURRENCY (they are sent to the client according to sniffs)
- Added new field in DB "EarnedQuantity" to store the total earned (this value isn't reset unlike TrackedQuantity).
- Implemented several currency flags:
  - CurrencyIgnoreMaxQtyOnLoad: Archaeology currencies don't send the max quantity in SMSG_SETUP_CURRENCY
  - CurrencyTrackQuantity & CurrencyBUseTotalEarnedForEarned: Use of the tracked quantity and earned quantity respectively
  - CurrencySuppressChatMessages & CurrencySuppressChatMessageOnVersionChange: Replacement of the parameter printLog
  - CurrencyDynamicMaximum: Used to calculate a maximum variable amount (CurrencyTypes.db2->MaxQtyWorldStateID & SPELL_EFFECT_INCREASE_CURRENCY_CAP)
- Added worldstates related to currencies
- Implemented SPELL_EFFECT_INCREASE_CURRENCY_CAP (field IncreasedCapQuantity in DB)
- Implemented SPELL_AURA_MOD_CURRENCY_CATEGORY_GAIN_PCT

**Issues addressed:**
None

**Tests performed:**
Builds, tested in-game.

**Known issues and TODO list:** (add/remove lines as needed)
- Player::RewardPlayerWithRewardPack can reward with currencies from different sources so a unique CurrencyGainSource cannot be used
- I don't know the CurrencyDestroyReason value which is used in retail for TraitCurrencyType::CurrencyTypesBased (Player::ApplyTraitEntryChanges)
- I don't know if CurrencyTypeFlags::ComputedWeeklyMaximum is currently in use for current exp

### CODE REVIEW

File: src/server/game/DataStores/DBCEnums.h [hunk 1]
```diff
@@ -664,6 +664,62 @@ enum class ChrCustomizationReqFlag : int32
...
 
+enum CurrencyConsts
+{
+    CURRENCY_TYPE_ANCIENT_MANA    = 1155,
+    CURRENCY_TYPE_AZERITE         = 1553,
+
+    CURRENCY_MAX_CAP_ANCIENT_MANA = 2000
+};
+
+enum CurrencyTypesFlags : uint32
```
Comment (MaxtorCoder) [inline src/server/game/DataStores/DBCEnums.h:675]:
```suggestion
enum class CurrencyTypesFlags : uint32
```
That way you can remove the Currency* prefixes from enum entries

File: src/server/game/DataStores/DBCEnums.h [hunk 2]
```diff
@@ -664,6 +664,62 @@ enum class ChrCustomizationReqFlag : int32
...
+    CurrencyHasWarmodeBonus                    = 0x08000000, // NYI
+    CurrencyIsAllianceOnly                     = 0x10000000,
+    CurrencyIsHordeOnly                        = 0x20000000,
+    CurrencyLimitWarmodeBonusOncePerTooltip    = 0x40000000, // NYI
+    DeprecatedCurrencyFlag                     = 0x80000000
+};
+
+DEFINE_ENUM_FLAG(CurrencyTypesFlags);
+
+enum CurrencyTypesFlagsB : uint32
```
Comment (MaxtorCoder) [inline src/server/game/DataStores/DBCEnums.h:713]:
```suggestion
enum class CurrencyTypesFlagsB : uint32
```
Same as CurrencyTypesFlags

**Review by MaxtorCoder — CHANGES_REQUESTED** [review]

File: src/server/game/DataStores/DBCEnums.h (hunk 1 above)
Comment (meji46) [reply src/server/game/DataStores/DBCEnums.h]: Thanks! Updated the enum to enum class

I've taken the names as they appear here: https://github.com/tomrus88/BlizzardInterfaceCode/blob/master/Interface/AddOns/Blizzard_APIDocumentationGenerated/CurrencyConstantsDocumentation.lua (using Currency* prefix)

File: src/server/game/Server/Packets/MiscPackets.cpp [hunk 3]
```diff
@@ -50,14 +50,19 @@ WorldPacket const* WorldPackets::Misc::SetCurrency::Write()
...
+
     _worldPacket.WriteBit(WeeklyQuantity.has_value());
     _worldPacket.WriteBit(TrackedQuantity.has_value());
     _worldPacket.WriteBit(MaxQuantity.has_value());
     _worldPacket.WriteBit(TotalEarned.has_value());
     _worldPacket.WriteBit(SuppressChatLog);
     _worldPacket.WriteBit(QuantityChange.has_value());
-    _worldPacket.WriteBit(QuantityLostSource.has_value());
     _worldPacket.WriteBit(QuantityGainSource.has_value());
+    _worldPacket.WriteBit(QuantityLostSource.has_value());
```
Comment (Shauren) [inline src/server/game/Server/Packets/MiscPackets.cpp:65]: no. lost then gain is correct

Comment (meji46) [reply src/server/game/Server/Packets/MiscPackets.cpp:65]: Renown reputation from level 9 to 10:
```
ServerToClient: SMSG_SET_CURRENCY (0x2574) Length: 30 ConnIdx: 1
Type: 2002
Quantity: 10
Flags: 0
UiEventToastCount: 0
HasWeeklyQuantity: False
HasTrackedQuantity: False
HasMaxQuantity: True
HasTotalEarned: False
SuppressChatLog: True
HasQuantityChange: True
HasQuantityLostSource: True
HasQuantityGainSource: False
HasFirstCraftOperationID: False
HasLastSpendTime: False
MaxQuantity: 25
QuantityChange: 1
QuantityLostSource: 55
```

**Pull request merged**